src/renderer/api/kube-watch-api.ts

```diff
@@ -1,201 +1,326 @@
+    this.log({
+      message: "connection check",
+      meta: { connected: this.isConnected },
+    });
 
-    const query = await this.getQuery();
+    if (this.isConnected) return;
+
+    return this.connect();
+  }
+
+  protected async connect(apis = this.apis) {
+    this.disconnect(); // close active connections first
+
+    if (!navigator.onLine || !apis.length) {
+      this.isConnected = false;
 
-    if (!this.activeApis.length || !query.api.length) {
       return;
     }
 
-    const apiUrl = `${apiPrefix}/watch?${stringify(query)}`;
+    this.log({
+      message: "Connecting",
+      meta: { apis }
+    });
 
-    this.evtSource = new EventSource(apiUrl);
-    this.evtSource.onmessage = this.onMessage;
-    this.evtSource.onerror = this.onError;
-    this.writeLog("CONNECTING", query.api);
-  }
+    try {
+      const requestId = ++this.requestId;
+      const abortController = new AbortController();
 
-  reconnect() {
-    if (!this.evtSource || this.evtSource.readyState !== EventSource.OPEN) {
-      this.reconnectAttempts = this.maxReconnectsOnError;
-      this.connect();
-    }
-  }
+      const request = await fetch(`${apiPrefix}/watch`, {
+        method: "POST",
+        body: JSON.stringify({ apis } as IWatchRoutePayload),
+        signal: abortController.signal,
+        headers: {
+          "content-type": "application/json"
+        }
+      });
 
-  protected disconnect() {
-    if (!this.evtSource) return;
-    this.evtSource.close();
-    this.evtSource.onmessage = null;
-    this.evtSource = null;
-  }
+      // request above is stale since new request-id has been issued
+      if (this.requestId !== requestId) {
+        abortController.abort();
 
-  protected onMessage(evt: MessageEvent) {
-    if (!evt.data) return;
-    const data = JSON.parse(evt.data);
+        return;
+      }
 
-    if ((data as IKubeWatchEvent).object) {
-      this.onData.emit(data);
-    } else {
-      this.onRouteEvent(data);
-    }
-  }
+      let jsonBuffer = "";
+      const stream = request.body.pipeThrough(new TextDecoderStream());
+      const reader = stream.getReader();
 
-  protected async onRouteEvent(event: IKubeWatchRouteEvent) {
-    if (event.type === "STREAM_END") {
-      this.disconnect();
-      const { apiBase, namespace } = KubeApi.parseApi(event.url);
-      const api = apiManager.getApi(apiBase);
-
-      if (api) {
-        try {
-          await api.refreshResourceVersion({ namespace });
-          this.reconnect();
-        } catch (error) {
-          console.error("failed to refresh resource version", error);
-
-          if (this.subscribers.size > 0) {
-            setTimeout(() => {
-              this.onRouteEvent(event);
-            }, 1000);
-          }
-        }
+      this.isConnected = true;
+      this.reader = reader;
+
+      while (true) {
+        const { done, value } = await reader.read();
+
+        if (done) break; // exit
+
+        const events = (jsonBuffer + value).split("\n");
```

Comment:
Good question. Have to double check, but I assume it probably escaped somehow with `JSON.stringify`?

Reply:
I don't think json has the concept of multiple "documents" in a single file unlike yaml

Reply:
Actually all good. Try this: 
```js
JSON.stringify(x={test: "multi\nline"}).includes("\n")
```
So it's always single line for receiving json from the stream (it's strinfigied in watch-route).

Reply:
I guess I meant inter-record new lines. Since it seems like the stream is a bunch of JSON records concatenated together with new lines. Thanks for pointing out that in `watch-route` we are not adding any new lines to the serialization.